runtime_scan/pkg/provider/docker/client.go

```diff
@@ -0,0 +1,452 @@
+				{
+					Type:   mount.TypeVolume,
+					Source: assetVolume,
+					Target: mountPointPath,
+				},
+			},
+		},
+		&network.NetworkingConfig{
+			EndpointsConfig: map[string]*network.EndpointSettings{
+				config.AssetScanID: {
+					NetworkID: networkID,
+				},
+			},
+		},
+		nil,
+		containerName,
+	)
+	if err != nil {
+		return "", fmt.Errorf("failed to create scan container: %w", err)
+	}
+
+	return containerResp.ID, nil
+}
+
+func (c *Client) getContainerIdFromName(ctx context.Context, containerName string) (string, error) {
+	containers, err := c.dockerClient.ContainerList(ctx, types.ContainerListOptions{
+		All:     true,
+		Filters: filters.NewArgs(filters.Arg("name", containerName)),
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to list containers: %w", err)
+	}
+	if len(containers) == 0 {
+		return "", fmt.Errorf("scan container not found: %w", err)
+	}
+	if len(containers) > 1 {
+		return "", fmt.Errorf("found more than one scan container: %w", err)
+	}
+	return containers[0].ID, nil
+}
+
+func (c *Client) getNetworkIdFromName(ctx context.Context, networkName string) (string, error) {
+	networks, err := c.dockerClient.NetworkList(ctx, types.NetworkListOptions{
+		Filters: filters.NewArgs(filters.Arg("name", networkName)),
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to list networks: %w", err)
+	}
+	if len(networks) == 0 {
+		return "", fmt.Errorf("scan network not found: %w", err)
+	}
+	if len(networks) > 1 {
+		return "", fmt.Errorf("found more than one scan network: %w", err)
+	}
+	return networks[0].ID, nil
+}
+
+func (c *Client) exportAsset(ctx context.Context, config *provider.ScanJobConfig) (io.ReadCloser, func(), error) {
+	logger := log.GetLoggerFromContextOrDiscard(ctx)
+
+	objectType, err := config.AssetInfo.ValueByDiscriminator()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get asset object type: %w", err)
+	}
+
+	switch value := objectType.(type) {
+	case models.ContainerInfo:
+		containerID := *value.Id
+		contents, err := c.dockerClient.ContainerExport(ctx, containerID)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to export container: %w", err)
+		}
+		return contents, nil, nil
+
+	case models.ContainerImageInfo:
+		// Create an ephemeral container to export asset
+		image := *value.Name
+		containerResp, err := c.dockerClient.ContainerCreate(
+			ctx,
+			&container.Config{Image: image},
+			nil,
+			nil,
+			nil,
+			"",
+		)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to create helper container: %w", err)
+		}
+
+		cleanup := func() {
+			err := c.dockerClient.ContainerRemove(ctx, containerResp.ID, types.ContainerRemoveOptions{Force: true})
+			if err != nil {
+				logger.Errorf("failed to remove helper container=%s: %v", containerResp.ID, err)
+			}
+		}
+
+		contents, err := c.dockerClient.ContainerExport(ctx, containerResp.ID)
+		if err != nil {
+			return nil, cleanup, fmt.Errorf("failed to export container: %w", err)
```

Comment:
This cleanup action will not be executed due to non-nil error checks where exportAsset is run. Instead, you can execute cleanup action and return nil action before returning an error.

```go
cleanup()
return nil, nil, fmt.Errorf("failed to export container: %w", err)
```
Ideally, cleanup action is only returned when no errors have occurred, otherwise it is handled internally.

Reply:
This behaviour should also be documented for `exportAsset` method:
* Returns artifact in the form of a ReadCloser stream, along with a cleanup function that needs to be run once the work with the artifact is done. Cleanup action is only returned on nil error, and it can also be nil to indicate that no cleanup action needs to be performed.